### ISSUE TITLE
CVMFS server failure on CentOS 9

### ISSUE THREAD

**Issue by razvanvirtan** [cvmfs/cvmfs#2990]
### Description
Creating a repo on CentOS9 prints an error message because cvmfs is not able to sign the repo. The repository is created, but unusable, because a SEG fault is generated at the first transaction.

### Steps to reproduce
On CentOS9, run:
`sudo cvmfs_server mkfs <your_repo_name>`
and then
`sudo cvmfs_server transaction <your_repo_name>`

### Relevant output#

**Creating the repo**
```
$ sudo cvmfs_server mkfs razvan.test.repo
[sudo] password for razvan: 
Owner of razvan.test.repo [root]: razvan
Creating Configuration Files... done
Creating CernVM-FS Master Key and Self-Signed Certificate... done
Creating CernVM-FS Server Infrastructure... done
Creating Backend Storage... done
Signing 30 day whitelist with master key... The command rsautl was deprecated in version 3.0. Use 'pkeyutl' instead.
done
Creating Initial Repository... the private key '/etc/cvmfs/keys/razvan.test.repo.key' doesn't seem to match certificate '/etc/cvmfs/keys/razvan.test.repo.crt' (error:03000098:digital envelope routines::invalid digest)
failed to init repo connection
fail! (cannot sign repo)
```

**Using the repo**
```
$ sudo cvmfs_server transaction razvan.test.repo
/bin/cvmfs_server: line 4006: 173386 Segmentation fault      (core dumped) $(__publish_cmd dbg) transaction $@
```

**Comment by jblomer** [comment]
A quick workaround is setting `OPENSSL_ENABLE_SHA1_SIGNATURES=1` at the beginning of the `cvmfs_server` script.  I _think_ that is ok because we don't use the certificate for TLS things but only as an envelope for the public key. But I need to look further into it. Perhaps the solution is to fix the creation of the certificate & key.

**Comment by jblomer** [comment]
More precisely, the problem is in creating the [signature of the manifest](https://github.com/cvmfs/cvmfs/blob/cvmfs-2.9.4/cvmfs/signature.cc#L752). Creating the signature with SHA-1 digest is disallowed unless `OPENSSL_ENABLE_SHA1_SIGNATURES=1`. Interestingly, verification of a SHA-1 signature is not affected.

While I think EL9 is right to flag the use of the API in this way, cvmfs is still not much affected by SHA-1 collision attacks because it is the _hash_ of the manifest that is signed, not the manifest itself.  Still, it's a good idea to update the signature algorithm.  For backwards compatibility, we should add another field to the manifest that specifies the digest type.

(EL9 unit tests do uncover the issue, we should check how this slipped through)

**Comment by jblomer** [comment]
A possible alternative is changing to a statically linked LibreSSL (which we anyway already ship for macOS) for the libcrypto routines used in cvmfs (hashes, RSA, AES). This would also address the fact that OpenSSL 3 deprecated the low-level crypto API, seemingly without replacement.

**Comment by radupopescu** [comment]
Using the same SSL implementation everywhere may not be a bad idea at all. We should also be careful to update the bundled LibreSSL when there are any relevant security updates.

**Comment by razvanvirtan** [comment]
I also think the second idea would work better. I've tried the `OPENSSL_ENABLE_SHA1_SIGNATURES=1` and `export OPENSSL_ENABLE_SHA1_SIGNATURES=1` fix and it only worked when the repo owner was set to root (unexpected, since the owner shell is created with `-m`).

**Comment by jblomer** [comment]
> Using the same SSL implementation everywhere may not be a bad idea at all. We should also be careful to update the bundled LibreSSL when there are any relevant security updates.

+1 on updating our version of LibreSSL. Note that for TLS connections, curl would still dynamically link against the system default SSL/TLS library. The static LibreSSL would only be used for cryptographic primitives from libcrypto, which is rarely subject to security fixes.

**Comment by jblomer** [comment]
> I also think the second idea would work better. I've tried the `OPENSSL_ENABLE_SHA1_SIGNATURES=1` and `export OPENSSL_ENABLE_SHA1_SIGNATURES=1` fix and it only worked when the repo owner was set to root (unexpected, since the owner shell is created with `-m`).

Good point!  We can set the environment variable in the C++ code, e.g. in the signature manager's initialization code.

**Comment by radupopescu** [comment]
I gave it some more thought and I'm unsure using two SSL libraries simultaneously is a good solution or a good thing in general. Having two libraries provide the same symbols will make the linking process more intricate and fragile.

Additionally, it may not make sense to adopt LibreSSL everywhere, as, at least on Linux, OpenSSL seems to be very active and has caught up.

What about implementing support for creating new signature types on systems where SHA1 is deprecated, since validating existing signatures is still possible.

**Comment by jblomer** [comment]
The linking issue could be a showstopper. Regarding OpenSSL, the actual problem is that OpenSSL 3.0 deprecates a bunch of public functions that we need and will need for a long time (to support older clients).

If the EL9 workaround does the trick, we are not in a rush. So perhaps as a first step let's get that working and keep thinking about a proper fix.

**Comment by jblomer** [comment]
Reopening this one as a reminder for the necessary follow-up fixes.

**Comment by jblomer** [comment]
More workaround code in #3006

**Comment by jblomer** [comment]
A proper fix might work like this: we encapsulate the cryptographic classes in cvmfs (`hash.cc`, `signature.cc`, `encrypt.cc`) in a shared library `libcvmfs_crypto.so`. When compiling the library, we can statically link LibreSSL and only export cvmfs classes. The cryptographic classes make use of some of the cvmfs infrastructure code. So we would need to create a second, base library `libcvmfs_util.so`. Server binaries (and later client binaries, requires harmonization of compile options) link against either one or both of them.

That's some restructuring but it does come with other advantages. It should reduce build time and install space (we could create a base package `cvmfs-libs`). And it makes it easier in the future to change the provider of cryptographic primitives.

**Comment by jblomer** [comment]
Properly fixed per #3024
